concordium-consensus/src/Concordium/Scheduler.hs

```diff
@@ -1101,158 +1100,172 @@ handleContractUpdateV1 originAddr istance checkAndGetSender transferAmount recei
+                               -- Execution of the contract might have changed our own state. If so, we need to resume in the new state, otherwise
+                               -- we can keep the old one.
+                               (lastModifiedIndex, newState) <- getCurrentContractInstanceState istance
+                               let stateChanged = lastModifiedIndex /= modificationIndex
+                               resumeState <- getRuntimeReprV1 newState
+                               newBalance <- getCurrentContractAmount Wasm.SV1 istance
+                               go (resumeEvent True:evs ++ interruptEvent:events) =<< runInterpreter (return . WasmV1.resumeReceiveFun rrdInterruptedConfig resumeState stateChanged newBalance WasmV1.Success Nothing)
+                        Just (InstanceInfoV1 targetInstance) -> do
+                          -- If this returns Right _ it is successful, and we pass this, and the returned return value
+                          -- to the caller.
+                          -- Otherwise we roll back all the changes and return the return value, and the error code to the caller.
+                          let rName = Wasm.uncheckedMakeReceiveName (instanceInitName (iiParameters targetInstance)) imcName
+                          withRollback (handleContractUpdateV1 originAddr targetInstance (checkAndGetBalanceInstanceV1 ownerAccount istance) imcAmount rName imcParam) >>= \case
+                             Left cer ->
+                               go (resumeEvent False:interruptEvent:events) =<< runInterpreter (return . WasmV1.resumeReceiveFun rrdInterruptedConfig rrdCurrentState False entryBalance (WasmV1.Error cer) (WasmV1.ccfToReturnValue cer))
+                             Right (rVal, callEvents) -> do
+                               (lastModifiedIndex, newState) <- getCurrentContractInstanceState istance
+                               let stateChanged = lastModifiedIndex /= modificationIndex
+                               resumeState <- getRuntimeReprV1 newState
+                               newBalance <- getCurrentContractAmount Wasm.SV1 istance
+                               go (resumeEvent True:callEvents ++ interruptEvent:events) =<< runInterpreter (return . WasmV1.resumeReceiveFun rrdInterruptedConfig resumeState stateChanged newBalance WasmV1.Success (Just rVal))
+                    WasmV1.Upgrade{..} -> do
+                        -- charge for base administrative cost.
+                        tickEnergy Cost.initializeContractInstanceBaseCost
+                        newModule <- getModuleInterfaces imuModRef
+                        case newModule of
+                          -- The module could not be found.
+                          Nothing -> go (resumeEvent False:interruptEvent:events) =<< runInterpreter (return . WasmV1.resumeReceiveFun rrdInterruptedConfig rrdCurrentState False entryBalance (WasmV1.UpgradeInvalidModuleRef imuModRef) Nothing)
```

Comment:
I presume there's no need for version checking here since Upgrade is not available in P4. It might be worth documenting why this branch (and others as relevant) won't happen in P4.

Reply:
I agree, will add a followup PR. Will see if adding a type to just make this irrelevant as well.

Reply:
Will add.